llama.cpp

```diff
@@ -6781,6 +6904,156 @@ struct llm_build_context {
+
+                // prepare convolution for all tokens in the batch with a self-overlapping view,
+                // shifting by one column each ... depth? ... with a window of d_conv columns.
+                // {(d_conv-1)+n_tok, d_inner} => {d_conv, d_inner, n_tok}
+                conv_x = ggml_view_3d(ctx0, conv_x, d_conv, d_inner, n_tok, conv_x_nb1, 1*ggml_element_size(conv_x), 0);
+
+                // perform convolution
+                // => {1, d_inner, n_tok}
+                x = ggml_sum_rows(ctx0, ggml_mul(ctx0, conv_x, model.layers[il].ssm_conv1d));
+                // => {d_inner, n_tok, 1}
+                x = ggml_permute(ctx0, x, 2, 0, 1, 3);
+
+                // bias
+                x = ggml_add(ctx0, x, model.layers[il].ssm_conv1d_b);
+
+                x = ggml_silu(ctx0, x);
+            }
+
+            // ssm
+            {
+                //  {d_inner, dt_rank + 2*d_state} * {d_inner, n_tok} => {dt_rank + 2*d_state, n_tok}
+                struct ggml_tensor * x_db = ggml_mul_mat(ctx0, model.layers[il].ssm_x, x);
+                // split
+                struct ggml_tensor * dt = ggml_view_2d(ctx0, x_db, dt_rank, n_tok, x_db->nb[1], 0);
+                struct ggml_tensor * B  = ggml_view_2d(ctx0, x_db, d_state, n_tok, x_db->nb[1], ggml_element_size(x_db)*dt_rank);
+                struct ggml_tensor * C  = ggml_view_2d(ctx0, x_db, d_state, n_tok, x_db->nb[1], ggml_element_size(x_db)*(dt_rank+d_state));
+
+                // {dt_rank, d_inner} * {dt_rank, n_tok} => {d_inner, n_tok}
+                dt = ggml_mul_mat(ctx0, model.layers[il].ssm_dt, dt);
+                dt = ggml_add(ctx0, dt, model.layers[il].ssm_dt_b);
+
+                // Custom operator to implement some of the optimizations
+                // described in the Annex D of the Mamba paper.
+                // TODO: maybe also optimize step 4 of the Speed section of Annex D (the mul_mat with C)
+                // => {d_state, d_inner, n_tok}
+                ssm_state = ggml_ssm_scan(ctx0, ssm_state, x, dt, model.layers[il].ssm_a, B);
+
+                // only store last state
+                ggml_build_forward_expand(gf,
+                    ggml_cpy(ctx0,
+                        ggml_view_2d(ctx0, ssm_state, d_state, d_inner, ssm_state->nb[1], (n_tok-1)*ssm_state->nb[2]),
+                        ggml_view_tensor(ctx0, kv_self.v_l[il])));
+
+                struct ggml_tensor * y;
+                if (n_tok == 1) {
+                    // row-wise dot product ("dn,n->d")
+                    // {d_state, d_inner} * {d_state, 1} => {d_inner, 1}
+                    y = ggml_mul_mat(ctx0, ssm_state, C);
+                } else {
+                    // {d_state, d_inner, n_tok} * {d_state, n_tok} => {d_inner, 1, n_tok}
+                    y = ggml_mul_mat(ctx0, ssm_state, ggml_permute(ctx0, C, 0, 2, 1, 3));
+                    // => {d_inner, n_tok}
+                    y = ggml_permute(ctx0, y, 0, 2, 1, 3);
+                }
```

Comment:
I had done this branch as a premature optimization. Since it was only there to avoid permutes (which are cheaper than I thought when writing this), it is/was relatively easy to remove the branch. Good to know regarding dynamically changing the compute graph.